packages/yarnpkg-core/sources/Project.ts

```diff
@@ -703,9 +699,166 @@ export class Project {
 
       originalPackages.set(originalPkg.locatorHash, originalPkg);
 
-      const pkg = this.configuration.normalizePackage(originalPkg);
+      let pkg = this.configuration.normalizePackage(originalPkg);
+
+      const prettyParent = () => {
+        if (pkgParent)
+          return structUtils.prettyLocator(this.configuration, pkgParent);
+
+        const workspaceName = workspace.manifest.name;
+
+        if (workspaceName)
+          return structUtils.prettyIdent(this.configuration, workspaceName);
+
+        return `workspace without name`;
+      };
+
+      if (pkg.variants) {
```

Comment:
I suspect you don't want to process its own variant when a workspace lists one (otherwise it's not clear what it would produce, since we can't make it "replace" itself 🤔).

```suggestion
      if (pkg.variants && pkgParent) {
```

Reply:
I see that's already a distinction you make a bit below, but then I don't understand why splitting the two. Shouldn't `variants` just be completely ignored if there isn't a parent?